src/report/report.c

```diff
@@ -0,0 +1,231 @@
+
+        (void) sd_varlink_set_userdata(link, context);
+
+        r = sd_varlink_set_relative_timeout(link, TIMEOUT_USEC);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set varlink timeout: %m");
+
+        r = sd_varlink_attach_event(link, event, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink connection to event loop: %m");
+
+        r = sd_varlink_bind_reply(link, metrics_on_query_reply);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        r = sd_varlink_observe(link, "io.systemd.Metrics.List", /* parameters= */ NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to issue io.systemd.Metrics.List call: %m");
+
+        *vl = TAKE_PTR(link);
+
+        return 0;
+}
+
+static void sd_varlink_unref_many(sd_varlink **array, size_t n) {
+        assert(array);
+
+        FOREACH_ARRAY(v, array, n)
+                sd_varlink_unref(*v);
+
+        free(array);
+}
+
+static int metrics_start_query(void) {
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        int r, c;
+
+        d = opendir(RUN_SYSTEMD_METRICS_PATH);
+        if (!d) {
+                if (errno == ENOENT)
+                        return log_error_errno(-ESRCH, "No metrics sockets found: %m");
+
+                return log_error_errno(-errno, "Failed to open directory: %m");
+        }
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = sd_event_set_signal_exit(event, true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
+
+        size_t n_varlinks = MAX_CONCURRENT_METRICS_SOCKETS;
+        sd_varlink **varlinks = new0(sd_varlink *, n_varlinks);
+        if (!varlinks)
+                return log_error_errno(-ENOMEM, "Failed to allocate varlinks array: %m");
+
+        CLEANUP_ARRAY(varlinks, n_varlinks, sd_varlink_unref_many);
+
+        Context context = {};
+        int i = 0;
+
+        FOREACH_DIRENT(de, d, return -errno) {
+                _cleanup_free_ char *p = NULL;
+
+                if (de->d_type != DT_SOCK && de->d_type != DT_UNKNOWN)
+                        continue;
+
+                p = path_join(RUN_SYSTEMD_METRICS_PATH, de->d_name);
+                if (!p)
+                        return log_error_errno(-ENOMEM, "Failed to allocate path: %m");
+
+                r = metrics_call(p, event, &varlinks[i], &context);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to connect to %s: %m", p);
+                        continue;
+                }
+
+                context.n_ref++;
+
+                if (++i >= MAX_CONCURRENT_METRICS_SOCKETS) {
+                        log_warning("Too many concurrent metrics sockets, skipping %s", p);
```

Comment:
log_warning("Too many concurrent metrics sockets, skipping %s", p);

a bit misleading because you break

log_warning("Too many concurrent metrics sockets, stop iterating");

seems a bit better. But this is just a taste question